agent/ebs/watcher_test.go

```diff
@@ -70,233 +70,318 @@ func newTestEBSWatcher(ctx context.Context, agentState dockerstate.TaskEngineSta
-	var wg sync.WaitGroup
-	wg.Add(1)
-	mockDiscoveryClient.EXPECT().ConfirmEBSVolumeIsAttached(taskresourcevolume.TestDeviceName, taskresourcevolume.TestVolumeId).
-		Do(func(deviceName, volumeID string) {
-			wg.Done()
-		}).
-		Return(taskresourcevolume.TestDeviceName, nil).
-		MinTimes(1)
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
 
-	watcher.HandleResourceAttachment(ebsAttachment1)
-	watcher.HandleResourceAttachment(ebsAttachment2)
+			ctx := context.Background()
+			taskEngineState := dockerstate.NewTaskEngineState()
+			mockDiscoveryClient := mock_ebs_discovery.NewMockEBSDiscovery(mockCtrl)
+			mockTaskEngine := mock_engine.NewMockTaskEngine(mockCtrl)
+			mockDaemonManager := mock_daemonmanager.NewMockDaemonManager(mockCtrl)
+			mockCsiClient := mock_csiclient.NewMockCSIClient(mockCtrl)
+			stateChangeChan := make(chan statechange.Event)
 
-	// Instead of starting the EBS watcher, we'll be mocking a tick of the EBS watcher's scan ticker.
-	// Otherwise, the watcher will continue to run forever and the test will panic.
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		pendingEBS := watcher.agentState.GetAllPendingEBSAttachmentsWithKey()
-		if len(pendingEBS) > 0 {
-			foundVolumes := apiebs.ScanEBSVolumes(pendingEBS, watcher.discoveryClient)
-			watcher.NotifyAttached(foundVolumes)
-		}
-	}()
+			if tc.setDaemonRunningExpectations != nil {
+				tc.setDaemonRunningExpectations(mockCtrl, mockTaskEngine, mockDaemonManager, stateChangeChan)
+			}
+			if tc.setCSIClientExpectations != nil {
+				tc.setCSIClientExpectations(mockCtrl, mockCsiClient)
+			}
+			if tc.setConfirmEBSVolumeIsAttached != nil {
+				tc.setConfirmEBSVolumeIsAttached(mockCtrl, mockDiscoveryClient)
+			}
 
-	wg.Wait()
+			watcher := newTestEBSWatcher(ctx, taskEngineState, mockDiscoveryClient, mockTaskEngine, mockCsiClient)
+			for _, ebsAttachment := range tc.attachments {
+				err := watcher.HandleEBSResourceAttachment(ebsAttachment)
+				if !tc.shouldHandle {
+					assert.Error(t, err)
+				} else {
+					assert.NoError(t, err)
+				}
+			}
 
-	assert.Len(t, taskEngineState.(*dockerstate.DockerTaskEngineState).GetAllEBSAttachments(), 1)
-	ebsAttachment, ok := taskEngineState.(*dockerstate.DockerTaskEngineState).GetEBSByVolumeId(taskresourcevolume.TestVolumeId)
-	assert.True(t, ok)
-	assert.True(t, ebsAttachment.IsAttached())
+			assert.Len(t, taskEngineState.(*dockerstate.DockerTaskEngineState).GetAllEBSAttachments(), tc.expectedNumSavedAttachments)
+
+			ebsAttachment, ok := taskEngineState.(*dockerstate.DockerTaskEngineState).GetEBSByVolumeId(taskresourcevolume.TestVolumeId)
+			if !tc.shouldTick {
+				assert.False(t, ok)
+			} else {
+				// Note: There are cases where we need to tick more than once.
+				// Ex: Tick 1 -> Create a CSI driver task if it doesn't exist/isn't running, Tick 2 -> Find/Validate/Stage EBS volume
```

Comment:
In future maybe we should have separate unit tests for `tick` and `HandleEBSResourceAttachment` methods. That should make the tests simpler I think.

Reply:
Yeah that's true, I originally wanted to test also test the general workflow as well. Especially in the test case where we already have an existing CSI task that's running or stopped.

Reply:
Yeah I understand the motivation. But setting up the mocks a certain way to make that happen is equivalent to directly calling the two methods with arguments for the different test cases. :)

Reply:
Yep yep completely agree! I'll add this as a TODO.

Reply:
(+1 I'd like to see this too.)